VCRUISE: Add V-Cruise engine

elasota opened the pull request:
This adds initial support for the V-Cruise engine used by Reah: Face the Unknown and Schizm: Mysterious Journey and initial detection for Reah.  (I actually need to figure out the detection some more because the GOG version and DVD version only ship English VO.)

This is very preliminary right now, right now it boots Reah to the intro cinematic and then gets stuck at the opening screen, I'm working on getting navigation implemented.

Documentation of its formats (and a decoder/encoder for the scripts) currently resides at https://github.com/elasota/vcr, everything is documented except for yet-to-be-understood script ops and the "morph" format which was added in Schizm.

The biggest change to core is the addition of cursor masks.  Reah uses a magnifying glass cursor where the lens part is inverted, unfortunately ScummVM only supports color key cursors.  This tries to add the groundwork (i.e. mode values and feature flags) for fully supporting the Windows and classic MacOS cursor formats, although only total inversion is supported in OpenGL right now.

digitall commented:
@elasota : One warning of missing gameID in the detection table entries:
````
In file included from engines/vcruise/detection.cpp:37:
./engines/vcruise/detection_tables.h:49:1: warning: missing initializer for member ‘VCruise::VCruiseGameDescription::gameID’ [-Wmissing-field-initializers]
 };
````

digitall commented:
@elasota : That gets rid of a fair amount of warnings. Only a few remain:
````
engines/vcruise/runtime.cpp: In member function ‘void VCruise::Runtime::scriptOpDisc1(ScriptArg_t)’:
engines/vcruise/runtime.cpp:781:22: warning: variable ‘stackArgs’ set but not used [-Wunused-but-set-variable]
         StackValue_t stackArgs[n];                                                                \
                      ^~~~~~~~~
engines/vcruise/runtime.cpp:972:9: note: in expansion of macro ‘TAKE_STACK’
         TAKE_STACK(1);
         ^~~~~~~~~~
engines/vcruise/runtime.cpp: In member function ‘void VCruise::Runtime::scriptOpDisc2(ScriptArg_t)’:
engines/vcruise/runtime.cpp:781:22: warning: variable ‘stackArgs’ set but not used [-Wunused-but-set-variable]
         StackValue_t stackArgs[n];                                                                \
                      ^~~~~~~~~
engines/vcruise/runtime.cpp:978:9: note: in expansion of macro ‘TAKE_STACK’
         TAKE_STACK(2);
         ^~~~~~~~~~
engines/vcruise/runtime.cpp: In member function ‘void VCruise::Runtime::scriptOpDisc3(ScriptArg_t)’:
engines/vcruise/runtime.cpp:781:22: warning: variable ‘stackArgs’ set but not used [-Wunused-but-set-variable]
         StackValue_t stackArgs[n];                                                                \
                      ^~~~~~~~~
engines/vcruise/runtime.cpp:984:9: note: in expansion of macro ‘TAKE_STACK’
         TAKE_STACK(3);
         ^~~~~~~~~~
````

sev- commented:
Before going further into the engine review, I'd like to ask you to split the cursor changes into a separate PR so that it can be reviewed and merged before the engine.

digitall commented:
@elasota : Now clean of warnings for me. Thanks for working on this. If you need playtesting, I have Reah and Schizm from CD.

digitall commented:
@elasota : Further note on my 6 CD version of Reah. Missing detection entry:
````
Matched game IDs for the vcruise engine: reah-win

  {"Reah.exe", 0, "77bc7f7819cdd443f52b193529138c87", 305664},

````
For reference, full file md5sum is:
````
md5sum -b Reah/Reah.exe 
c8cce9a47fba364b9ec288e2e76bcc77 *Reah/Reah.exe
````

elasota commented:
> Before going further into the engine review, I'd like to ask you to split the cursor changes into a separate PR so that it can be reviewed and merged before the engine.

Moved this to pull #4713 but also fixed a bunch of issues, so will have to redo this PR after that's merged.

> Further note on my 6 CD version of Reah. Missing detection entry:

One thing I would be interested with the 6CD version is what format the music is in.  The DVD version uses MP3-in-WAV music but based on some things I've been reading the CD-ROM version may used uncompressed, so it would be nice to get an MD5 + file size of Music-00.wav too.

I am also not sure if there exist versions with other language VO.  The DVD version only has English VO.

digitall commented:
@elasota : Here are the required detection entries for my copies of Reah (6 CD) and Schizm (5 CD):
````
	{ // Reah: Face the Unknown, 6 CD Version
		{
			"reah",
			"CD",
			{
				{"Reah.exe", 0, "77bc7f7819cdd443f52b193529138c87", 305664},
				AD_LISTEND
			},
			Common::EN_ANY,
			Common::kPlatformWindows,
			ADGF_UNSTABLE,
			GUIO0()
		},
		GID_REAH,
	},
	{ // Schizm, 5 CD Version
		{
			"schizm",
			"CD",
			{
				{"Schizm.exe", 0, "296edd26d951c3bdc4d303c4c88b27cd", 364544},
				AD_LISTEND
			},
			Common::EN_ANY,
			Common::kPlatformWindows,
			ADGF_UNSTABLE,
			GUIO0()
		},
		GID_SCHIZM,
	},
````

digitall commented:
@elasota : My copy of Reah has Music-00.wav in pcm_s16le 2ch 22050Hz format i.e. uncompressed.
File md5sum is af80fad4ec51b3467b43f4ca531b7788 *Music-00.wav with size of 14395864 bytes

digitall commented:
@elasota : I can provide full file listing with MD5sum if required of both. Just a further note that currently when I try to start either, it immediately exits with a SDL_BlitSurface failed error:
````
Running Reah: Face the Unknown (CD/Windows/English)
Reah.exe: 77bc7f7819cdd443f52b193529138c87, 305664 bytes.
Surface::fillRect: bytesPerPixel must be 1, 2, or 4!
SDL_BlitSurface failed: Surfaces must not be locked during blit!
Debugger started, type 'exit' to return to the game.
Type 'help' to see a little list of commands and variables.
ERROR: SDL_BlitSurface failed: Surfaces must not be locked during blit!
Read 19 history entries
SDL_BlitSurface failed: Surfaces must not be locked during blit!
````

elasota commented:
It looks like the problem there is the graphics mode selection in SurfaceSDL can select a 24-bit mode and fillScreen with a 24-bit mode crashes SurfaceSDL.

elasota commented:
Removed the cursor changes, fixed the crash in SurfaceSDL.

digitall commented:
@elasota : That has fixed the crash for me. I get the Reah intro now up to the journalist outside of the walled city gate. I note that you didn't include the detection entries I provided yet :) Apart from that, the Schizm.exe is identical between the GOG release and the original 5 CD release so that may need another file in detection to distinguish if needed. You probably already know, but Schizm fails to start with "Couldn't figure out what screen to start on". Anyway, good luck on getting this working.

digitall commented:
@elasota : Just a quick note against commit e7b0634, this warning turned up:
engines/vcruise/runtime.cpp: In member function ‘void VCruise::Runtime::continuePlayingAnimation(bool, bool&)’:
engines/vcruise/runtime.cpp:346:50: warning: comparison of integer expressions of different signedness: ‘uint’ {aka ‘unsigned int’} and ‘int’ [-Wsign-compare]
                         if (_animDisplayingFrame >= static_cast<int>(_animLastFrame)) {

I would just change the type of one of those variables to match i.e. both signed or both unsigned instead and avoid the cast...

sev- commented on engines/vcruise/script.cpp:
```diff
@@ -0,0 +1,767 @@
+		return true;
+	}
+
+	if (token.size() >= 5 && token[0] == 'C' && token[1] == 'U' && token[2] == 'R' && token[3] == '_') {
```
`token.hasPrefix("CURS_")`

sev- commented on engines/vcruise/textparser.cpp:
```diff
@@ -0,0 +1,322 @@
+	memset(_returnedBuffer, 0, kReturnBufferSize);
+}
+
+bool TextParser::readOneChar(char &outC, TextParserState &outState) {
```
I'd recommend avoid using `char` and use `byte` instead, because `char` could be signed.

elasota replied:
I was considering that but it didn't seem like a good idea when the output of all the parser functions are `Common::String` which uses char.  `TextParser::isWhitespace` already accounts for the possibility of extended characters being negative and I don't think it matters in any other situation.

sev- commented on engines/vcruise/textparser.cpp:
```diff
@@ -0,0 +1,322 @@
+}
+
+bool TextParser::readOneChar(char &outC, TextParserState &outState) {
+	if (_returnBufferPos == kReturnBufferSize) {
```
I recommend using `wrapBufferedReadStream()` to achieve the same result. See `common/bufferedstream.h`

sev- replied:
Was this lost?

elasota replied:
It no longer buffers input, but it does buffer returned (requeued) characters, which is necessary because this was implemented with requeue instead of lookahead.

sev- commented on engines/vcruise/textparser.cpp:
```diff
@@ -0,0 +1,322 @@
+	outInt = static_cast<int16>(i);
+}
+
+void TextParser::expectInt(int &outInt, const Common::String &blamePath) {
```
sscanf() has all of this logic already

sev- commented on engines/vcruise/textparser.cpp:
```diff
@@ -0,0 +1,322 @@
+	outUInt = result;
+}
+
+void TextParser::expectLine(Common::String &outToken, const Common::String &blamePath, bool continueToNextLine) {
```
consider `Common::String::readLine()`

elasota replied:
I think this is preferable because it's easier to keep up line/col error diagnostics synced up.

sev- commented on engines/vcruise/vcruise.cpp:
```diff
@@ -0,0 +1,187 @@
+
+VCruiseEngine::VCruiseEngine(OSystem *syst, const VCruiseGameDescription *gameDesc) : Engine(syst), _gameDescription(gameDesc) {
+	const Common::FSNode gameDataDir(ConfMan.get("path"));
+}
```
You need to call `SearchMan.addDirectory(_gameDataDir.getPath(), _gameDataDir);`, so the SearchMan could work

sev- commented on engines/vcruise/vcruise.cpp:
```diff
@@ -0,0 +1,187 @@
+		videoSize = Common::Point(608, 348);
+		menuBarSize = Common::Point(640, 44);
+		traySize = Common::Point(640, 88);
+		exeName = "Reah.exe";
```
why not use `_gameDescription->desc.filesDescriptions[0].fileName`?

sev- commented on engines/vcruise/vcruise.cpp:
```diff
@@ -0,0 +1,187 @@
+}
+
+Common::Error VCruiseEngine::run() {
+	Common::List<Graphics::PixelFormat> pixelFormats = _system->getSupportedFormats();
```
I'd recommend using `Engine::initGraphicsModes(const Graphics::ModeList &modes)` instead of this whole construct

elasota replied:
`initGraphicsModes` is just for sizes, and I don't like sending a mode list because `findCompatibleFormat` requires exact mode matches when really I don't care about channel order or alpha channel presence as long as a mode with enough RGB bits is available.

sev- commented on engines/vcruise/vcruise.cpp:
```diff
@@ -0,0 +1,187 @@
+bool VCruiseEngine::hasFeature(EngineFeature f) const {
+	switch (f) {
+	case kSupportsReturnToLauncher:
+	case kSupportsSavingDuringRuntime:
```
I see no save support yet. This needs to be disabled for the time being so the GMM does not get confused.

sev- commented on engines/vcruise/vcruise.h:
```diff
@@ -0,0 +1,79 @@
+ *
+ */
+
+#ifndef VCRUISE_VCRUISE_H
```
Please use just `VCRUISE_H` here

sev- reviewed (changes requested):
Finished my first review round

elasota commented on engines/vcruise/runtime.cpp:
```diff
@@ -0,0 +1,1396 @@
+
+		Common::Rect copyRect = Common::Rect(0, 0, surface->w, surface->h);
+
+		Common::Rect constraintRect = Common::Rect(0, 0, _gameSection.rect.width(), _gameSection.rect.height());
```
Yes, but this is a stub for future feature support.  The way this engine works is that animations are assigned to frame ranges of a single large fullscreen-size AVI, but some of those animations copy only part of the AVI to the framebuffer to allow independently-animated parts.  An easy example is the very first puzzle, with the gate bars.  However, that feature is not supported yet, so right now `copyRect` is always the entire frame.

elasota commented on engines/vcruise/detection.cpp:
```diff
@@ -0,0 +1,105 @@
+			game.appendGUIOptions(Common::getGameGUIOptionsDescriptionLanguage(Common::PL_POL));
+			game.appendGUIOptions(Common::getGameGUIOptionsDescriptionLanguage(Common::ES_ESP));
+			game.appendGUIOptions(Common::getGameGUIOptionsDescriptionLanguage(Common::EL_GRC));
+			game.appendGUIOptions(Common::getGameGUIOptionsDescriptionLanguage(Common::RU_RUS));
```
The only releases I'm aware of are multi-lingual.  I will adjust this in the future if we get new information.

elasota commented:
I think all issues should be cleaned up now except for the pixel format and `expectLine` implementation which I think there are valid technical reasons for doing the way they are, and the AVI `copyRect` handling which is intentional.

sev- commented on engines/vcruise/runtime.cpp:
```diff
@@ -0,0 +1,1342 @@
+					if (stream->read(interactionData, 12) != 12)
+						error("Error reading interaction data");
+
+					idef.rect = Common::Rect(READ_LE_INT16(interactionData + 0), READ_LE_INT16(interactionData + 2), READ_LE_INT16(interactionData + 4), READ_LE_INT16(interactionData + 6));
```
Why not use here a more intuitive `stream->readUint16LE()` for every variable?

elasota replied:
C++ order of parameter evaluation is undefined (which I have been burned by before - VS optimizer can reorder execution of parameter expressions even if the expressions have side effects) so it would need to be separate read lines, plus I think this way makes it easier to detect a read failure.

sev- reviewed (approved):
Whoa! You're fast. I have only one small issue/remark which could be resolved in-tree.

sev- commented on engines/vcruise/textparser.cpp:
```diff
@@ -0,0 +1,269 @@
+}
+
+bool TextParser::readOneChar(char &outC, TextParserState &outState) {
+	if (_returnedBufferPos == kReturnedBufferSize) {
```
I recommend using `wrapBufferedReadStream()` to achieve the same result. See `common/bufferedstream.h`

sev- commented:
Time to merge...
